build.gradle.kts

```diff
@@ -422,9 +424,86 @@ subprojects {
+                val configuration: Configuration = configurations.detachedConfiguration(
+                        dependencies.create(depModule)
+                )
+                baselineArtifact = files(configuration.files).filter {
+                    it.name.equals(depJar)
+                }.singleFile
+            } finally {
+                project.group = oldGroup
+            }
+            if (newVersion != null) {
+                try {
+                    project.group = "virtual_new_for_japicmp"
+                    val depModule = "io.opentelemetry:${base.archivesBaseName}:$newVersion@jar"
+                    val depJar = "${base.archivesBaseName}-${newVersion}.jar"
+                    val configuration: Configuration = configurations.detachedConfiguration(
+                            dependencies.create(depModule)
+                    )
+                    newArtifact = files(configuration.files).filter {
+                        it.name.equals(depJar)
+                    }.singleFile
+                } finally {
+                    project.group = oldGroup
+                }
+            }
+
+            tasks {
+                val jar = getByName("jar") as Jar
+                if (newArtifact == null) {
+                    newArtifact = file(jar.archiveFile)
+                }
+                create<JapicmpTask>("japicmp") {
+                    dependsOn("jar")
+                    oldClasspath = files(baselineArtifact)
+                    newClasspath = files(newArtifact)
+                    htmlOutputFile = file("$buildDir/reports/japi.html")
+                    //only output changes, not everything
+                    isOnlyModified = true
+                    //this is needed so that we only consider the current artifact, and not dependencies
+                    isIgnoreMissingClasses = true
+                    // double wildcards don't seem to work here (*.internal.*)
+                    packageExcludes = listOf("*.internal", "io.opentelemetry.internal.shaded.jctools.*")
+                    if (newVersion == null) {
+                        val baseVersionString = if (userRequestedBase == null) "latest" else baselineVersion
+                        txtOutputFile = file("$projectDir/docs/api_diff_current_vs_${baseVersionString}.txt")
+                    } else {
+                        txtOutputFile = file("$projectDir/docs/api_diff_${newVersion}_vs_${baselineVersion}.txt")
+                    }
+                }
+                //have the build task depend on the api comparison task, to make it more likely it will get used.
+                getByName("build").dependsOn("japicmp")
+            }
+        }
+    }
+
     plugins.withId("maven-publish") {
+        //generate the api diff report for any module that is stable and publishes a jar.
+        if (!project.hasProperty("otel.release")
+                && project.tasks.map { it.name }.contains("jar")) {
```

Comment:
Do you think it's simpler to just do something like `project.name.startsWith("bom")` instead of the jar task check? While in practice we would never do it, but I try to avoid ordering dependencies where possible, checking for `jar` means the java plugin must have been applied before the maven-publish plugin in the build script, fine in practice, but if can be avoided it's sort of nice.

Reply:
good question. I don't have a strong opinion on this at all. I was just trying to get something working. :)

Reply:
yep, this works fine and is definitely cleaner.